### ISSUE TITLE
Next challenge ... My FITS images are in the RGGB pixel format from an astrophotography camera (ASI, using a Sony color chip).

### ISSUE THREAD

**Issue by brampf** [brampf/fitscore#10]
Next challenge ... My FITS images are in the RGGB pixel format from an astrophotography camera (ASI, using a Sony color chip).

1) I like to convert the FITS with either CoreImage or with CVPixelBufferCreate... to a raw image in order to apply all the nice RawFilters from iOS/Swift to them.

I'm completely stuck. I can create a pixel buffer with the kCVPixelFormaType_14Bayer_RGGB, but the quickview in XCode only shows me a grayscale image. I am not able to convert it to a CIImage or CGImage showing colors, it is either nil or remains a grayscale image that is not de-bayered / demosaiced.

2) Looking at your RGB conversion of the FITS (FITSBytesTool), I noticed the RGB are consecutively ordered (3x layersize). This is not the case for me, where I have a Bayer matrix. Re-writing this function to separate the pixel colors, does result in a color image, but the colors are all wrong...

Any hint to solve either problem would be highly appreciated.

_Originally posted by @Planet30 in https://github.com/brampf/fitscore/discussions/8#discussioncomment-346832_

**Comment by brampf** [comment]
Just created a new issue for this inquiry as the other is already marked as answered and I almost didn't see it there ;)

In fact, I did not yet consider RGGB files so far. As I said, FITSKit is pretty basic at the moment. Maybe this is the time to work on this for a change. 

Could you share a sample image and / or the spec of the sensor? That way, I'll give it a look and guide or even help you with an implementation.

**Comment by Planet30** [comment]
Thank you for answering so quickly and creating the "issue".

Following your "warning", I actually do not use the FITSkit anymore directly, I rather copy/pasted (more or less 1:1) what from your functions into my own functions.

I see two options:
1) Creating a CIImage in raw format and let iOS/Swift do the developing.  I could not get this to work....
2) Creating an RGB image directly from the RGGB pixels.

Here is the link to the FITS image:
https://www.icloud.com/iclouddrive/0K4SxrqGaPtY6_HlQwPdqv-OA#M42


And here its header info:

--- PrimaryHDU : 23388736 ------------------------ 
Header:  0 - 2720 padded to 2880
Data:    2880 - 23391616 padded to 23394240
Optional(SIMPLE  =                    T / file does conform to FITS standard             
BITPIX  =                   16 / number of bits per data pixel                  
NAXIS   =                    2 / number of data axes                            
NAXIS1  =                 4144 / length of data axis 1                          
NAXIS2  =                 2822 / length of data axis 2                          
EXTEND  =                    T / FITS dataset may contain extensions            
COMMENT                     FITS (Flexible Image Transport System) format is defined in 'Astronomy
COMMENT                     and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H
BZERO   =                32768 / physical = BZERO + BSCALE*array_value          
BSCALE  =                    1 / physical = BZERO + BSCALE*array_value          
XBINNING=                    1 / Binning factor in width                        
YBINNING=                    1 / Binning factor in height                       
EXPOINUS=             10000000 / Exposure time in us                            
GAIN    =                  120 / The ratio of output / input                    
OFFSET  =                   30 / Brightness(offset) of image                    
WB_RED  =                   50 / White Balance(Red)                             
WB_BLUE =                   50 / White Balance(Blue)                            
CBLACK  =                    0 / Initial display black level in ADUs            
CWHITE  =                65535 / Initial display white level in ADUs            
PEDESTAL=                    0 / Correction to add for zero-based ADU           
SWCREATE= 'ASIImg  '           / Name of software that created the image        
SWOWNER = 'ZWO     '           / Licensed owner of software                     
DATE-OBS= -01-15T22:11:38.310' / UTC start date of observation                  
BAYERPAT= 'RGGB    '           / Debayer pattern,such as RGGB,BGGR,GRBG,GBRG    
COLORTYP= 'RAW16   '           / Color space, such as RAW8,RAW16,RGB24          
INPUTFMT= 'FITS    '           / Format of file from which image was read       
CSTRETCH= 'Medium  '           / Initial display stretch mode                   
INSTRUME= 'ZWO ASI294MC Pro'   / Camera model                                   
XPIXSZ  =                 4.63 / Pixel Width in um                              
YPIXSZ  =                 4.63 / Pixel Height in um                             
EXPTIME =                 10.0 / Total Exposure Time (s)                        
EXPOSURE=                 10.0 / Exposure time in seconds                       
CCD-TEMP=                  0.0 / CMOS sensor temperature in C                   
END                                                                             
)

**Comment by brampf** [comment]
You only got 2 Axis there. With RGB you would usually get `NAXIS = 3` and with RGGB I was expecting something like `NAXIS = 4`. Decoding this HDU as an 4144x2822 pixel with 16bit per "pixel" (which is exactly 4144*2822*2 = 23391616 bytes) does only contain one single layer and thus can only be interpreted as one single grayscale image. 

This could mean either one of two things:

1. A look at the sensor spec (https://www.sony-semicon.co.jp/products/common/pdf/IMX294CJK_Flyer.pdf) gives me the idea that you might need to demoniac this into basically a 4x4 bit image before re-assembling as RGGB and rendering it. 
2. There are three more IMAGE extensions in that FITS file, treating each layer of the final color image as a separate grayscale image. In that case you would need to assemble those into one color image. But the first case seems more likely ;)

**Comment by Planet30** [comment]
This is exactly the question: How to do the de-Bayer or de-mosaic of the RGGB pixel layout. This is very common for all DSLR cameras and there are libraries in OpenCV, Python and even with in iOS/SWIFT to "develop" the raw image into an RGB image. I just need to get the FITS image data into an array or pixel buffer format, which will make iOS interpret it as one of the 670 known camera raw pixel formats.

**Comment by brampf** [comment]
since this is almost a month, I wanted to let you know that I'm still working on this. It took me a while until the information that this sensor is not just RGGB bug quad binned RGGB. I read this information in the file but did not understand what this means exactly until today.

I'm currently studying various algorithms on who to demoniac this the binary data and compute the RGB picture from it. It may take another while but I'll figure this one out ;)

**Comment by Planet30** [comment]
Hello, Thanks for keeping this up :-)
There are open libraries in python and opencv for demosaicing.
BUT ... I have been trying to let iOS do this faster and presumably better than the open-source libraries. The internal camera sensors of the iDevices do also happen to have a CFA (color filter array) with an RGGB layout. There is a raw mode in the core video and core image frameworks to handle these raw data automatically. Either with CV-create-pixelbuffer or via CIFilter it should be possible. CIFilter would have the immense advantage of being able to use all the CIRawFilter methods in iOS/SWIFT on the actual raw data, which would provide much higher quality images than the "manual" demosaicing. I have not be able yet, however, to figure it out myself... I keep trying on my side.

**Comment by brampf** [comment]
From what I have seen so far "core image" and "automatically" are to words which only hardly go into the same sentence. As I said, your sensor in fact is not RGGB but "quad binned" RGGB which I have not seen in any of apples documentation. But I don't haven't done a lot with core image so far as I was primarily focused on accelerate to process raw data.
Unfortunately when I research for using core image to demosaic images, I end up on https://stackoverflow.com/questions/66417989/cvpixelbuffer-to-cgimage-conversion-only-gives-black-white-image whos' author just seems a little familiar ;)

**Comment by brampf** [comment]
I finally pushed the advanced branch in 2eabf51 with some experimental color coding of RGGB CFA as well as my first attempt of demosaicing with bilinear interpolation of the missing color values. All in all it's not very performant but definitely going in the right direction.
This branch requires the 'fitskit-advanced' branch b40d3e0  of fitscore to compile.

**Comment by Planet30** [comment]
After some more digging, I found dcraw, which is exactly made for debayering/demosaicing of some 600 RAW color image formats from all major camera brands (NIKON, CANON, ...).

Unfortunately, FITS/RGGB is not one of them. dcraw is kind of not maintained anymore. It is written in C and can be easily compiled on UNIX/LINUX/macOS. Since I started off with SWIFT, and never did any programming in C, I would not take up on the task of adding FITS to dcraw. I am sure there are other much more capable people out there than me to do this task quickly and professionally.

Below two links for dcraw and one website using it in combination with imageMagick for debayering RGGB raw files:

dcraw for macOS:
https://www.empiricalimaging.com/knowledge-base/compile-dcraw-for-macos/

Demosaicing with imageMagick and dcraw:
http://im.snibgo.com/demosaic.htm

NOTE:
This shows again that there is no straight forward way to do this on iOS/LINUX and SWIFT. There are more and more "inexpensive" astrophotography color cameras coming on the market that only export raw format as FITS/RGGB.

**Comment by brampf** [comment]
I spend some time in researching demosaicing algorithms and am going to implement Variable Number of Gradients (VNG) and maybe even Aliasing Minimization and Zipper Elimination (AMaZE).

Generally speaking, these do not look overly complicated in theory but come at a high cost in terms of computation required per pixel. I therefore currently play around on how to improve performance on the bilinear interpolation wich already works pretty well by itself but is currently terrible slow. And this is - in theory - the least expensive algorithm for demosaicing. But I'm sure I can improve that.

BTW. writing a SPM wrapper around a C/ C++ library in order to use it with a swift application isn't an overly complicated thing to do. Handling different licenses on the other hand - sometimes - is.

My personal goal here is to gain experience with image processing in general and writing a pure, native Swift library. Otherwise, I would have simply written a SPM wrapper package for cfitsio instead of fitscore in the first place :D

**Comment by Planet30** [comment]
I fully agree! I just did not want to put an extra burden on you here by being too demanding!

Same here, I want to learn how to handle all of this in SWIFT and finally have something for my iPad out in the field.

BTW, today I found the program ASTAP, which actually does exactly what we had discussed above... using cfitsio, dcraw/libraw. It also offers different de-bayering methods. While it seems rather powerful, it is very slow. I think this could be improved when utilising different Apple/SWIFT frameworks such as Accelerate (vForce, vDSP, vImage) and Core Image/Graphics and maybe even Metal. I found vDSP very fast on an iPad Pro.

**Comment by krissbennett** [comment]
Hi folks, I'm really interested in this too, I'd like to render the FITS from my ZWO camera to preview them, and I have managed to load the FITS file ok, but its really dark and not showing much of the detail. Do you have any examples on how to adjust properties to get it to display better? I am trying to use a sample file from my ZWO ASI 294MC Pro which is RGGB one shot color astronomy camera. 

Thanks, any further guidance in the right direction or code samples would be great :) 

Thanks

**Comment by Planet30** [comment]
Hi, welcome to the club. I also have the ASI294MC. We are working on this. This FITS conversation will not do the automatic histogram correction. This you have to program yourself. This is what I have originally intended by using Core Image and CIFilter for raw image conversion and processing.  Not straight forward. ... but we will get there eventually.

**Comment by krissbennett** [comment]
Thanks, would love to help if I can, if you have any further progress you can point me at that I can look at, I don't claim to know the solution but would love to help solve it if I can be of any use.

**Comment by brampf** [comment]
I just pushed 113c12e to the advanced-branch with a working ProofOfConcept for an performance-optimised version of the bilinear algorithm. Check it out and simply call
`       let file = try FitsFile.read(contentsOf: url)!
        
        let converter = Demosaic_Bilinear_GCD<Int16>(parameter: .RGGB)
        
        let image = try file.prime.render(converter)`

**Comment by Planet30** [comment]
Thank you :-) !!! Will try it out this week!

**Comment by brampf** [comment]
Earlier today I generalised the approach and provided a protocol to implement future decoding algorithms besides the not-as-sophisticated bilinear interpolation. To demosaic your picture, it's now simply
`let image : CGImage? = try file.prime.decode(Demosaic_Bilinear_GCD.self, .RGGB)` 
This code renders the image in under a second on my MacBook 13" 2020.

**Comment by krissbennett** [comment]
Hi thanks that works for me, and my image seems to load well (although can't quite tell if loading in color), just need to figure out the Automatic histogram so I can see a preview like ASIFitsViewer :)
Is there settings you'd recommend to tweak for this?

I've attached link to my fits file and screenshots for reference. 

https://we.tl/t-AUUmuhhFWQ
Thanks again for this tool, will be great to be able to view my fits files on my iPad :)

**Comment by Planet30** [comment]
Hi. Yes, it works :-) And pretty fast on my iPad Pro. Great work.

Colours are off... naturally, because the white balancing is always done in the post processing of the hard- and software of the device incorporating the imaging chip.

I started looking into different color filters of core image. It works in real-time easily, but adjusting the image to the right colours, brightness, etc. can be tricky.
https://developer.apple.com/library/archive/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html#//apple_ref/doc/filter/ci/CIGammaAdjust

For the histogram stretching, I am planning to look into Metal Performance Shaders:
https://developer.apple.com/documentation/metalperformanceshaders/image_filters
It offers different histogram equalisers (with "z" ;-).

**Comment by brampf** [comment]
when I started working on fitscore / fitskit, I also started working on an iPad app and made some prototypes for color correction using curves using apples accelerate framework. This is in fact why I'm decoding the files always as ARGBFFFF because accelerate expects an alpha layer when working with floating point pixels - which I chose as a universal format since floating point operations have a lot more performance on hardware than integer operations.

Long story short, give me a few weeks to "reactivate" that project and update from pre-big-sur to SwiftUI 2.0 and I'll send you and TestFlight invitation. I guess it would make a lot of sense to join forces in that direction instead of us three developing three different iPad apps.

**Comment by brampf** [comment]
Before moving to iPad project, I felt the urge to clean up and make the new code of the advanced branch available on main. Therefore, I created a pull request. maybe you could do me the honor of a review?
brampf/fitskit/pull/3
